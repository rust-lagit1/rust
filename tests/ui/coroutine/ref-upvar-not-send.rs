// For `Send` coroutines, suggest a `T: Sync` requirement for `&T` upvars,
// and suggest a `T: Send` requirement for `&mut T` upvars.

#![feature(coroutines, stmt_expr_attributes)]

fn assert_send<T: Send>(_: T) {}
//~^ NOTE required by a bound in `assert_send`
//~| NOTE required by this bound in `assert_send`
//~| NOTE required by a bound in `assert_send`
//~| NOTE required by this bound in `assert_send`
//~| NOTE required by a bound in `assert_send`
//~| NOTE required by this bound in `assert_send`
//~| NOTE required by a bound in `assert_send`
//~| NOTE required by this bound in `assert_send`

fn main() {
    let x: &*mut () = &std::ptr::null_mut();
    //~^ NOTE has type `&*mut ()` which is not `Send`
    let y: &mut *mut () = &mut std::ptr::null_mut();
    //~^ NOTE has type `&mut *mut ()` which is not `Send`
    assert_send(
        //~^ ERROR coroutine cannot be sent between threads safely
        //~| NOTE coroutine is not `Send`
        #[coroutine]
        move || {
            //~^ ERROR coroutine cannot be sent between threads safely
            //~| NOTE coroutine is not `Send`
            //~| NOTE coroutine is not `Send`
            //~| yield occurs here
            yield;
            let _x = x;
            //~^ NOTE captured value is not `Send`
            //~| NOTE has type `&*mut ()` which is not `Send`, because `*mut ()` is not `Sync`
        },
    );
    assert_send(
        //~^ ERROR coroutine cannot be sent between threads safely
        //~| NOTE coroutine is not `Send`
        #[coroutine]
        move || {
            //~^ ERROR coroutine cannot be sent between threads safely
            //~| NOTE coroutine is not `Send`
            //~| NOTE coroutine is not `Send`
            //~| yield occurs here
            yield;
            let _y = y;
            //~^ NOTE captured value is not `Send`
            //~| NOTE has type `&mut *mut ()` which is not `Send`, because `*mut ()` is not `Send`
        },
    );
}
