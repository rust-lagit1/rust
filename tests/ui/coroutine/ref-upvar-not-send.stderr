error: coroutine cannot be sent between threads safely
  --> $DIR/ref-upvar-not-send.rs:25:9
   |
LL | /         move || {
LL | |
LL | |
LL | |
...  |
LL | |
LL | |         },
   | |_________^ coroutine is not `Send`
   |
   = help: the trait `Sync` is not implemented for `*mut ()`, which is required by `{coroutine@$DIR/ref-upvar-not-send.rs:25:9: 25:16}: Send`
note: captured value is not `Send` because `&` references cannot be sent unless their referent is `Sync`
  --> $DIR/ref-upvar-not-send.rs:31:22
   |
LL |             let _x = x;
   |                      ^ has type `&*mut ()` which is not `Send`, because `*mut ()` is not `Sync`
note: required by a bound in `assert_send`
  --> $DIR/ref-upvar-not-send.rs:6:19
   |
LL | fn assert_send<T: Send>(_: T) {}
   |                   ^^^^ required by this bound in `assert_send`

error: coroutine cannot be sent between threads safely
  --> $DIR/ref-upvar-not-send.rs:40:9
   |
LL | /         move || {
LL | |
LL | |
LL | |
...  |
LL | |
LL | |         },
   | |_________^ coroutine is not `Send`
   |
   = help: within `{coroutine@$DIR/ref-upvar-not-send.rs:40:9: 40:16}`, the trait `Send` is not implemented for `*mut ()`, which is required by `{coroutine@$DIR/ref-upvar-not-send.rs:40:9: 40:16}: Send`
note: captured value is not `Send` because `&mut` references cannot be sent unless their referent is `Send`
  --> $DIR/ref-upvar-not-send.rs:46:22
   |
LL |             let _y = y;
   |                      ^ has type `&mut *mut ()` which is not `Send`, because `*mut ()` is not `Send`
note: required by a bound in `assert_send`
  --> $DIR/ref-upvar-not-send.rs:6:19
   |
LL | fn assert_send<T: Send>(_: T) {}
   |                   ^^^^ required by this bound in `assert_send`

error: coroutine cannot be sent between threads safely
  --> $DIR/ref-upvar-not-send.rs:21:5
   |
LL | /     assert_send(
LL | |
LL | |
LL | |         #[coroutine]
...  |
LL | |         },
LL | |     );
   | |_____^ coroutine is not `Send`
   |
   = help: the trait `Sync` is not implemented for `*mut ()`, which is required by `{coroutine@$DIR/ref-upvar-not-send.rs:25:9: 25:16}: Send`
note: coroutine is not `Send` as this value is used across a yield
  --> $DIR/ref-upvar-not-send.rs:25:9
   |
LL |     let x: &*mut () = &std::ptr::null_mut();
   |         - has type `&*mut ()` which is not `Send`
...
LL |         move || {
   |         ^ yield occurs here, with `x` maybe used later
note: required by a bound in `assert_send`
  --> $DIR/ref-upvar-not-send.rs:6:19
   |
LL | fn assert_send<T: Send>(_: T) {}
   |                   ^^^^ required by this bound in `assert_send`

error: coroutine cannot be sent between threads safely
  --> $DIR/ref-upvar-not-send.rs:36:5
   |
LL | /     assert_send(
LL | |
LL | |
LL | |         #[coroutine]
...  |
LL | |         },
LL | |     );
   | |_____^ coroutine is not `Send`
   |
   = help: within `{coroutine@$DIR/ref-upvar-not-send.rs:40:9: 40:16}`, the trait `Send` is not implemented for `*mut ()`, which is required by `{coroutine@$DIR/ref-upvar-not-send.rs:40:9: 40:16}: Send`
note: coroutine is not `Send` as this value is used across a yield
  --> $DIR/ref-upvar-not-send.rs:40:9
   |
LL |     let y: &mut *mut () = &mut std::ptr::null_mut();
   |         - has type `&mut *mut ()` which is not `Send`
...
LL |         move || {
   |         ^ yield occurs here, with `y` maybe used later
note: required by a bound in `assert_send`
  --> $DIR/ref-upvar-not-send.rs:6:19
   |
LL | fn assert_send<T: Send>(_: T) {}
   |                   ^^^^ required by this bound in `assert_send`

error: aborting due to 4 previous errors

